# Data Preprocessing

In [40]:
# importing libraries
from utils.DataIngestion import get_path, get_config

import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
from sklearn.preprocessing import StandardScaler
import os

In [2]:
# getting the path
path = get_path()
print(f"Path: {path}")

Path: D:\Machine Learning\MACHINE LEARNING PROJECTS\airbnb_price_prediction\data


In [3]:
# getting the config and file information
config = get_config("data_preparation_config.yml")
file = config['file_names']['pickled_output_dataframe']
print(f"File: {file}")

Path, path_to_yaml: D:\Machine Learning\MACHINE LEARNING PROJECTS\airbnb_price_prediction\notebooks\data_preparation_config.yml
File: AB_NYC_2019_output_nov8.pkl


In [59]:
# reading the data
df = pd.read_pickle(os.path.join(path, file))
df.reset_index(inplace=True)
print(df.shape)
df.head()

(44977, 10)


index neighbourhood_group neighbourhood        room_type  price  \
0      0            Brooklyn    Kensington     Private room    149   
1      1           Manhattan       Midtown  Entire home/apt    225   
2      2           Manhattan        Harlem     Private room    150   
3      3            Brooklyn  Clinton Hill  Entire home/apt     89   
4      4           Manhattan   East Harlem  Entire home/apt     80   

   minimum_nights  number_of_reviews  reviews_per_month  \
0               1                  9               0.21   
1               1                 45               0.38   
2               3                  0               0.00   
3               1                270               4.64   
4              10                  9               0.10   

   calculated_host_listings_count  availability_365  
0                               6               365  
1                               2               355  
2                               1               365  
3                               1               194  
4                               1                 0

In [60]:
# null values:
df.isnull().sum()

index                             0
neighbourhood_group               0
neighbourhood                     0
room_type                         0
price                             0
minimum_nights                    0
number_of_reviews                 0
reviews_per_month                 0
calculated_host_listings_count    0
availability_365                  0
dtype: int64

In [61]:
# dtypes
df.dtypes

index                               int64
neighbourhood_group                object
neighbourhood                      object
room_type                          object
price                               int64
minimum_nights                      int64
number_of_reviews                   int64
reviews_per_month                 float64
calculated_host_listings_count      int64
availability_365                    int64
dtype: object

In [62]:
# features list 
list(df)

['index',
 'neighbourhood_group',
 'neighbourhood',
 'room_type',
 'price',
 'minimum_nights',
 'number_of_reviews',
 'reviews_per_month',
 'calculated_host_listings_count',
 'availability_365']

In [63]:
# independent and dependent features
dependent_feat = ['price']
independent_feat = ['neighbourhood_group', 'neighbourhood','room_type', 'minimum_nights', 'number_of_reviews', 
                    'reviews_per_month', 'calculated_host_listings_count', 'availability_365']

In [64]:
# continuous and categorical features:
categorical_feat = ['neighbourhood_group', 'room_type', 'neighbourhood']
cont_feat = ['minimum_nights', 'number_of_reviews', 'reviews_per_month', 'calculated_host_listings_count', 'availability_365']

In [65]:
# Number of unique categories in categorical features:
print(f"Neighbourhood Group: {df['neighbourhood_group'].nunique()} values")
print(f"Neighbourhoods: {df['neighbourhood'].nunique()} values")
print(f"Room Type: {df['room_type'].nunique()} values")

Neighbourhood Group: 5 values
Neighbourhoods: 219 values
Room Type: 3 values


#### Dependent Feature:

In [66]:
y = df[['price']]

#### Encoding the Categorical Features

- First -> encoding with the neighbourhood feature.
- Second -> encoding without the neighourhood feature as there are 219 unique values.

In [67]:
encoded_with = pd.get_dummies(df[categorical_feat])
encoded_without = pd.get_dummies(df[['neighbourhood_group', 'room_type']])

In [68]:
encoded_with.head()

neighbourhood_group_Bronx  neighbourhood_group_Brooklyn  \
0                          0                             1   
1                          0                             0   
2                          0                             0   
3                          0                             1   
4                          0                             0   

   neighbourhood_group_Manhattan  neighbourhood_group_Queens  \
0                              0                           0   
1                              1                           0   
2                              1                           0   
3                              0                           0   
4                              1                           0   

   neighbourhood_group_Staten Island  room_type_Entire home/apt  \
0                                  0                          0   
1                                  0                          1   
2                                  0                          0   
3                                  0                          1   
4                                  0                          1   

   room_type_Private room  room_type_Shared room  neighbourhood_Allerton  \
0                       1                      0                       0   
1                       0                      0                       0   
2                       1                      0                       0   
3                       0                      0                       0   
4                       0                      0                       0   

   neighbourhood_Arden Heights  ...  neighbourhood_Westchester Square  \
0                            0  ...                                 0   
1                            0  ...                                 0   
2                            0  ...                                 0   
3                            0  ...                                 0   
4                            0  ...                                 0   

   neighbourhood_Westerleigh  neighbourhood_Whitestone  \
0                          0                         0   
1                          0                         0   
2                          0                         0   
3                          0                         0   
4                          0                         0   

   neighbourhood_Williamsbridge  neighbourhood_Williamsburg  \
0                             0                           0   
1                             0                           0   
2                             0                           0   
3                             0                           0   
4                             0                           0   

   neighbourhood_Willowbrook  neighbourhood_Windsor Terrace  \
0                          0                              0   
1                          0                              0   
2                          0                              0   
3                          0                              0   
4                          0                              0   

   neighbourhood_Woodhaven  neighbourhood_Woodlawn  neighbourhood_Woodside  
0                        0                       0                       0  
1                        0                       0                       0  
2                        0                       0                       0  
3                        0                       0                       0  
4                        0                       0                       0  

[5 rows x 227 columns]

In [69]:
encoded_without.head()

neighbourhood_group_Bronx  neighbourhood_group_Brooklyn  \
0                          0                             1   
1                          0                             0   
2                          0                             0   
3                          0                             1   
4                          0                             0   

   neighbourhood_group_Manhattan  neighbourhood_group_Queens  \
0                              0                           0   
1                              1                           0   
2                              1                           0   
3                              0                           0   
4                              1                           0   

   neighbourhood_group_Staten Island  room_type_Entire home/apt  \
0                                  0                          0   
1                                  0                          1   
2                                  0                          0   
3                                  0                          1   
4                                  0                          1   

   room_type_Private room  room_type_Shared room  
0                       1                      0  
1                       0                      0  
2                       1                      0  
3                       0                      0  
4                       0                      0

#### Numerical Features

In [70]:
# list of numerical features
cont_feat

['minimum_nights',
 'number_of_reviews',
 'reviews_per_month',
 'calculated_host_listings_count',
 'availability_365']

In [72]:
cont_df = df[cont_feat]
print(cont_df.shape)
cont_df.head()

(44977, 5)


minimum_nights  number_of_reviews  reviews_per_month  \
0               1                  9               0.21   
1               1                 45               0.38   
2               3                  0               0.00   
3               1                270               4.64   
4              10                  9               0.10   

   calculated_host_listings_count  availability_365  
0                               6               365  
1                               2               355  
2                               1               365  
3                               1               194  
4                               1                 0

*Scaling the data:*

In [73]:
scaler = StandardScaler()
cont_df = pd.DataFrame(scaler.fit_transform(cont_df), columns=cont_feat)
cont_df.tail()

minimum_nights  number_of_reviews  reviews_per_month  \
44972       -0.452185          -0.530167          -0.688475   
44973       -0.219526          -0.530167          -0.688475   
44974        0.478453          -0.530167          -0.688475   
44975       -0.568515          -0.530167          -0.688475   
44976        0.129464          -0.530167          -0.688475   

       calculated_host_listings_count  availability_365  
44972                       -0.145030         -0.766197  
44973                       -0.145030         -0.558565  
44974                       -0.178543         -0.627776  
44975                       -0.010976         -0.820027  
44976                       -0.178543         -0.658536

In [74]:
cont_df.shape

(44977, 5)

#### Combining the Numerical and Categorical Features

In [75]:
X_with_neighbourhood = pd.concat([encoded_with, cont_df], axis=1)
X_without_neighbourhood = pd.concat([encoded_without, cont_df], axis=1)

In [76]:
print(X_with_neighbourhood.shape)
X_with_neighbourhood.head()

(44977, 232)


neighbourhood_group_Bronx  neighbourhood_group_Brooklyn  \
0                          0                             1   
1                          0                             0   
2                          0                             0   
3                          0                             1   
4                          0                             0   

   neighbourhood_group_Manhattan  neighbourhood_group_Queens  \
0                              0                           0   
1                              1                           0   
2                              1                           0   
3                              0                           0   
4                              1                           0   

   neighbourhood_group_Staten Island  room_type_Entire home/apt  \
0                                  0                          0   
1                                  0                          1   
2                                  0                          0   
3                                  0                          1   
4                                  0                          1   

   room_type_Private room  room_type_Shared room  neighbourhood_Allerton  \
0                       1                      0                       0   
1                       0                      0                       0   
2                       1                      0                       0   
3                       0                      0                       0   
4                       0                      0                       0   

   neighbourhood_Arden Heights  ...  neighbourhood_Willowbrook  \
0                            0  ...                          0   
1                            0  ...                          0   
2                            0  ...                          0   
3                            0  ...                          0   
4                            0  ...                          0   

   neighbourhood_Windsor Terrace  neighbourhood_Woodhaven  \
0                              0                        0   
1                              0                        0   
2                              0                        0   
3                              0                        0   
4                              0                        0   

   neighbourhood_Woodlawn  neighbourhood_Woodside  minimum_nights  \
0                       0                       0       -0.568515   
1                       0                       0       -0.568515   
2                       0                       0       -0.335855   
3                       0                       0       -0.568515   
4                       0                       0        0.478453   

   number_of_reviews  reviews_per_month  calculated_host_listings_count  \
0          -0.332158          -0.558263                       -0.010976   
1           0.459877          -0.452854                       -0.145030   
2          -0.530167          -0.688475                       -0.178543   
3           5.410101           2.188590                       -0.178543   
4          -0.332158          -0.626470                       -0.178543   

   availability_365  
0          1.971470  
1          1.894569  
2          1.971470  
3          0.656467  
4         -0.835408  

[5 rows x 232 columns]

In [77]:
print(X_without_neighbourhood.shape)
X_without_neighbourhood.head()

(44977, 13)


neighbourhood_group_Bronx  neighbourhood_group_Brooklyn  \
0                          0                             1   
1                          0                             0   
2                          0                             0   
3                          0                             1   
4                          0                             0   

   neighbourhood_group_Manhattan  neighbourhood_group_Queens  \
0                              0                           0   
1                              1                           0   
2                              1                           0   
3                              0                           0   
4                              1                           0   

   neighbourhood_group_Staten Island  room_type_Entire home/apt  \
0                                  0                          0   
1                                  0                          1   
2                                  0                          0   
3                                  0                          1   
4                                  0                          1   

   room_type_Private room  room_type_Shared room  minimum_nights  \
0                       1                      0       -0.568515   
1                       0                      0       -0.568515   
2                       1                      0       -0.335855   
3                       0                      0       -0.568515   
4                       0                      0        0.478453   

   number_of_reviews  reviews_per_month  calculated_host_listings_count  \
0          -0.332158          -0.558263                       -0.010976   
1           0.459877          -0.452854                       -0.145030   
2          -0.530167          -0.688475                       -0.178543   
3           5.410101           2.188590                       -0.178543   
4          -0.332158          -0.626470                       -0.178543   

   availability_365  
0          1.971470  
1          1.894569  
2          1.971470  
3          0.656467  
4         -0.835408

*Final Datasets with Independent and Dependent features:*

In [78]:
df_without_neighbourhood = pd.concat([X_without_neighbourhood, y], axis=1)
df_with_neighbourhood = pd.concat([X_with_neighbourhood, y], axis=1)

In [79]:
print(df_with_neighbourhood.shape)
print(df_without_neighbourhood.shape)

(44977, 233)
(44977, 14)


In [80]:
# saving the dataframes as CSV
print("Saving the Processed Data Files..")
df_with_neighbourhood.to_pickle(os.path.join(path, "final_with_neighbourhood.pkl"))
df_without_neighbourhood.to_pickle(os.path.join(path, "final_without_neighbourhood.pkl"))

Saving the Processed Data Files..
